common/ManagementTestShared/ManagementRecordTestBase.cs

```diff
@@ -0,0 +1,181 @@
+        /// </summary>
+        protected abstract void InitializeClients();
+
+        /// <summary>
+        /// Setup of resources that will be run before a test fixture or suite
+        /// This should be overriden if resources need to be setup once for a suite of test
+        /// </summary>
+        protected virtual Task OnOneTimeSetupAsync()
+        {
+            this.InitializeClients();
+            return Task.FromResult<object>(null);
+        }
+
+        /// <summary>
+        /// Setup of resources that will run before each test in a test fixture or suite
+        /// This should be overriden if resources need to be setup before each test.
+        /// </summary>
+        protected virtual Task OnSetupAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        /// <summary>
+        /// Teardown of resources that will be run after all test in a test fixture or suite have run
+        /// This should be overriden if customization for resource teardown needs done, else the teardown of
+        /// resources will be handled automaticall by the framework
+        /// </summary>
+        protected virtual Task OnOneTimeTearDownAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        /// <summary>
+        /// Teardown of resources that will be run after each test in test fixture or suite
+        /// This should be overriden if customization for resource teardown needs done, else the teardown of
+        /// resources will be handled automaticall by the framework
+        /// </summary>
+        protected virtual Task OnTearDownAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        /// <summary>
+        /// Called by NUnit for one time setup, calls on setup
+        /// </summary>
+        [OneTimeSetUp]
+        protected async Task RunOneTimeSetup()
+        {
+            if (Mode == RecordedTestMode.Live || Debugger.IsAttached)
+            {
+                Logger = new TestLogger();
+            }
+            Recording = new TestRecording(Mode, GetSessionFilePath(), Sanitizer, Matcher);
+            TestEnvironment.SetRecording(Recording);
+            InitializeClients();
+            await OnOneTimeSetupAsync();
+        }
+        /// <summary>
+        /// Called by NUnit to run one time tear down and runs user defined on one time teardown
+        /// </summary>
+        [OneTimeTearDown]
+        protected async Task RunOneTimeTearDown()
+        {
+            await StopTestRecording();
+            await CleanupResourceGroupsAsync();
+            Logger?.Dispose();
+            Logger = null;
+            await OnOneTimeTearDownAsync();
+        }
+        /// <summary>
+        /// A method called by NUnit framework to setup recordings for new test and run onsetup
+        /// </summary>
+        [SetUp]
+        protected async Task StartTestRecording()
+        {
+            await StopTestRecording();
```

Comment:
This is because we need to record and save what happens in the one time setup; I will move this to the last line in RunOneTimeSetup since that is the proper spot for it.

Reply:
Perhaps exposing "Save" and calling that directly is more appropriate than stop?

Similar to Flushing a stream without closing the stream?